omod/src/main/java/org/openmrs/module/pihcore/fragment/controller/dashboardwidgets/MedsDispensedFragmentController.java

```diff
@@ -0,0 +1,172 @@
+                patient = patientService.getPatient((Integer)patientConfig);
+            }
+            else if (patientConfig instanceof String) {
+                patient = patientService.getPatientByUuid((String)patientConfig);
+            }
+        }
+        if (patient == null) {
+            throw new IllegalArgumentException("No patient found. Please pass patient into the configuration");
+        }
+        model.put("patient", patient);
+        model.put("app", app);
+
+        // The existing behavior seems to be
+        // Get the last 5 encounters of the given type (PihEmrConfigConstants.ENCOUNTERTYPE_MEDICATION_DISPENSED_UUID)
+        // Render one row for each observation in these encounters with concept = CustomAppLoaderConstants.MED_DISPENSED_NAME_UUID
+        // If any of the encounters has no observations with this concept, render an empty row
+        // Where the observation has a valueDrug, display the associated Concept (using short name if available)
+
+        Map<String, List<Concept>> medsDispensed = new HashMap<>();
+        DateFormat df = new SimpleDateFormat("yyyy-MM-dd");
+        Locale locale = Context.getLocale();
+
+        String encounterTypesStr = getConfigValue(app, "encounterTypes");
+        List<EncounterType> encounterTypes = new ArrayList<>();
+        if (StringUtils.isNotBlank(encounterTypesStr)) {
+            for (String type : encounterTypesStr.split(",")) {
+                encounterTypes.add(encounterService.getEncounterTypeByUuid(type.trim()));
+            }
+        }
+
+        // Get all Med Dispensed Name observations from the given encounters
+        Concept medDispensedName = conceptService.getConceptByUuid(CustomAppLoaderConstants.MED_DISPENSED_NAME_UUID);
+        List<Obs> obsList = obsService.getObservationsByPersonAndConcept(patient, medDispensedName);
+        for (Obs o : obsList) {
+            Encounter e = o.getEncounter();
+            if (encounterTypes.isEmpty() || (e != null && encounterTypes.contains(e.getEncounterType()))) {
+                Concept concept = (o.getValueDrug() != null ? o.getValueDrug().getConcept() : o.getValueCoded());
+                Date dispenseDate = e.getEncounterDatetime();
+                List<Concept> medsOnDate = medsDispensed.computeIfAbsent(df.format(dispenseDate), k -> new ArrayList<>());
+                medsOnDate.add(concept);
+            }
+        }
+
+        // Get all Med Dispensed records from the Domain
+        MedicationDispenseCriteria criteria = new MedicationDispenseCriteria();
+        criteria.setPatient(patient);
+        criteria.setIncludeVoided(false);
+        List<MedicationDispense> dispenses = medicationDispenseService.getMedicationDispenseByCriteria(criteria);
+        for (MedicationDispense d : dispenses) {
+            Concept medConcept = (d.getDrug() != null ? d.getDrug().getConcept() : d.getConcept());
+            Date dispenseDate = d.getDateHandedOver();
+            if (dispenseDate == null) {
```

Comment:
I need to check the workflow again, but it's possible to have a dispense row for a drug that is "on hold" or for a drug that is "prepared" but not handed over (or that is long-term intent) so I believe that we only want to include dispenses with a "date_handed_over" (assuming the business logic is only to display those drugs the patient has actually received).  

You should be able to look at the new Dispense app to confirm how it works (or I can remind myself later)

Reply:
Oh, I see, you are only checking the encounter... I don't think we associate dispensing with encounters, so this may be fine.

Reply:
OK, I'll change.  I didn't know off-hand what fields the dispensing esm was populating and when.  If I can always rely upon the "dateHandedOver" being populated when a medication has actually been dispensed, I'll just stick with that, and if it is null I won't include it.  Make sense @mogoodrich ?